miniapps/adjoint/advection_diffusion.cpp

```diff
@@ -0,0 +1,488 @@
+// systems: ADAMS method and functional iteration.
+// It uses scalar relative and absolute tolerances.
+
+// In addition to the solution, sensitivities with respect to p1
+// and p2 as well as with respect to initial conditions are
+// computed for the quantity:
+//    g(t, u, p) = int_x u(x,t) at t = 5
+// These sensitivities are obtained by solving the adjoint system:
+//    dv/dt = -p1 * d^2 v / dx^2 + p2 * dv / dx
+// with homogeneous Ditrichlet boundary conditions and the final
+// condition:
+//    v(x,t=5) = 1.0
+// Then, v(x, t=0) represents the sensitivity of g(5) with respect
+// to u(x, t=0) and the gradient of g(5) with respect to p1, p2 is
+//    (dg/dp)^T = [  int_t int_x (v * d^2u / dx^2) dx dt ]
+//                [  int_t int_x (v * du / dx) dx dt     ]
+
+// This version uses MPI for user routines.
+// Execute with Number of Processors = N,  with 1 <= N <= MX.
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+#include <algorithm>
+
+#ifndef MFEM_USE_SUNDIALS
+#error This example requires that MFEM is built with MFEM_USE_SUNDIALS=YES
+#endif
+
+using namespace std;
+using namespace mfem;
+
+// Implement the adjoint rate equations in AdvDiffSUNDIALS
+class AdvDiffSUNDIALS : public TimeDependentAdjointOperator
+{
+public:
+   AdvDiffSUNDIALS(int ydot_dim, int ybdot_dim, Vector p,
+                   ParFiniteElementSpace *fes, Array<int> & ess_tdof) :
+      TimeDependentAdjointOperator(ydot_dim, ybdot_dim),
+      p_(p),
+      M(NULL), K(NULL), K_adj(NULL),
+      Mf(NULL), MK(NULL),
+      m(NULL), k(NULL),
+      pfes(fes),
+      M_solver(fes->GetComm()),
+      ess_tdof_list(ess_tdof)
+   {
+      int skip_zeros = 0;
+      ParMesh * pmesh = pfes->GetParMesh();
+
+      cout << "Essential tdofs: " << endl;
+      ess_tdof_list.Print();
+
+      m = new ParBilinearForm(pfes);
+      m->AddDomainIntegrator(new MassIntegrator());
+      m->Assemble(skip_zeros);
+      m->Finalize(skip_zeros);
+
+      k = new ParBilinearForm(pfes);
+      k->AddDomainIntegrator(
+	   new DiffusionIntegrator(
+		  *(new ConstantCoefficient(-p_[0]))));
+      Vector p2(fes->GetParMesh()->SpaceDimension());
+      p2 = p_[1];
+      k->AddDomainIntegrator(
+	   new ConvectionIntegrator(
+		  *(new VectorConstantCoefficient(p2))));
+      k->Assemble(skip_zeros);
+      k->Finalize(skip_zeros);
+
+      k1 = new ParBilinearForm(pfes);
+      k1->AddDomainIntegrator(
+	   new DiffusionIntegrator(
+		  *(new ConstantCoefficient(p_[0]))));
+      k1->AddDomainIntegrator(
+	   new ConvectionIntegrator(
+		  *(new VectorConstantCoefficient(p2))));
+      k1->Assemble(skip_zeros);
+      k1->Finalize(skip_zeros);
+
+      M = m->ParallelAssemble();
+      M->EliminateRowsCols(ess_tdof_list);
+
+      K = k->ParallelAssemble();
+      K->EliminateRowsCols(ess_tdof_list);
+
+      K_adj = k1->ParallelAssemble();
+      K_adj->EliminateRowsCols(ess_tdof_list);
+
+      MK = ParMult(M, K);
```

Comment:
Correct me if I'm wrong but you are eliminating all rows and cols corresponding to essential dofs before in M and K. This results in a 1 on the diagonal on M and in K and therefore you have a 2 on the diagonal.

This matrix is also not used anywhere.

Reply:
I'm not sure about the analysis of M*K, but it's not used anywhere.